File: examples/python/demo_probe.py
#!/usr/bin/env python3
"""
mp4decrypt-api - Media Probing & Metadata Extraction Demo (Python)
==================================================================

Demonstrates inspecting MP4 / fMP4 containers to extract track details,
codecs, dimensions, sample rates, encryption schemes, and default KIDs
without decrypting the file.

Run with:
    python examples/python/demo_probe.py
"""

import os
import sys
import urllib.request

# Ensure sdk/python is importable
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..")))
from sdk.python.mp4decrypt_sdk import Mp4DecryptSession


def main():
    print("=" * 67)
    print("      MP4DECRYPT-API - MEDIA CONTAINER PROBING (PYTHON)            ")
    print("=" * 67)

    # 1. Query dynamic library version
    ver = Mp4DecryptSession.get_version()
    print(f"\nEngine Version:    {ver.version_string}")
    print(f"ABI Revision:      {ver.abi_revision}")
    print(f"Bento4 Core:       {ver.bento4_version}")
    print(f"Build Date:        {ver.build_date}")

    # 2. Fetch sample media
    base_url = "https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey"
    print("\nDownloading sample init fragment...")
    init_bytes = urllib.request.urlopen(f"{base_url}/15/init.mp4").read()

    # 3. Probe in-memory buffer
    print(f"\nProbing {len(init_bytes):,} bytes from memory buffer:")
    info = Mp4DecryptSession.probe(init_bytes)

    print(f"  Container Brands:   {info.container_brands}")
    print(f"  Container Duration: {info.duration_seconds:.2f} s")
    print(f"  Is Fragmented:      {info.is_fragmented}")
    print(f"  Total Tracks:       {info.track_count}")

    for idx, track in enumerate(info.tracks):
        print(f"\n  --- Track #{idx + 1} (ID: {track.track_id}) ---")
        print(f"    Stream Type:      {track.stream_type} (handler: '{track.handler_type}')")
        print(f"    Codec:            {track.codec}")
        print(f"    Duration:         {track.duration_seconds:.2f} s")
        print(f"    Encrypted:        {track.is_encrypted}")
        if track.is_encrypted:
            print(f"    Scheme Type:      {track.scheme_type}")
            print(f"    Default KID:      {track.default_kid_hex}")
        if track.stream_type == "Audio":
            print(f"    Audio Channels:   {track.audio_channels}")
            print(f"    Sample Rate:      {track.sample_rate} Hz")
        elif track.stream_type == "Video":
            print(f"    Resolution:       {track.width}x{track.height}")
            print(f"    Framerate:        {track.frame_rate:.2f} fps")

    print("\nSUCCESS: Media probed and detailed metadata retrieved!")


if __name__ == "__main__":
    main()

File: examples/bun/demo_probe.ts
/**
 * mp4decrypt-api - Media Probing Demo (Bun)
 * =========================================
 *
 * Demonstrates inspecting container format, video resolution, framerate,
 * audio channels, codecs, protection schemes, and default KIDs without
 * requiring decryption keys.
 *
 * Run with:
 *   bun examples/bun/demo_probe.ts
 */

import { Mp4DecryptSession } from "../../sdk/bun/mp4decrypt_sdk";
import fs from "node:fs";
import path from "node:path";

async function main() {
  console.log("===================================================================");
  console.log("         MP4DECRYPT-API - MEDIA PROBING DEMO (BUN)                 ");
  console.log("===================================================================");

  console.log(`API Version    : ${Mp4DecryptSession.apiVersion}`);
  console.log(`Version String : ${Mp4DecryptSession.versionString}`);

  const baseUrl = "https://media.axprod.net/TestVectors/v7-MultiDRM-SingleKey";
  console.log(`\nFetching test vector media from ${baseUrl}/15/init.mp4...`);

  const initResp = await fetch(`${baseUrl}/15/init.mp4`);
  if (!initResp.ok) {
    throw new Error(`Failed to download test vector: ${initResp.status} ${initResp.statusText}`);
  }
  const initBuf = new Uint8Array(await initResp.arrayBuffer());
  console.log(`Downloaded ${initBuf.byteLength} bytes.`);

  console.log("\n[1] Probing in-memory buffer with Mp4DecryptSession.probe(buffer)...");
  const info = Mp4DecryptSession.probe(initBuf);

  console.log("\nContainer Information:");
  console.log(`  Brands        : ${info.containerBrands}`);
  console.log(`  Duration      : ${info.durationSeconds.toFixed(2)}s`);
  console.log(`  Is Fragmented : ${info.isFragmented ? "Yes (fMP4)" : "No"}`);
  console.log(`  Track Count   : ${info.trackCount}`);

  console.log("\nTracks:");
  for (const track of info.tracks) {
    console.log(`  - Track #${track.trackId} [${track.streamType}]`);
    console.log(`      Handler Type  : ${track.handlerType}`);
    console.log(`      Codec         : ${track.codec}`);
    console.log(`      Encrypted     : ${track.isEncrypted ? "YES" : "NO"}`);
    if (track.isEncrypted) {
      console.log(`      Scheme Type   : ${track.schemeType}`);
      console.log(`      Default KID   : ${track.defaultKidHex}`);
    }
    if (track.streamType === "VIDEO") {
      console.log(`      Resolution    : ${track.width}x${track.height}`);
      console.log(`      Frame Rate    : ${track.frameRate.toFixed(2)} fps`);
    } else if (track.streamType === "AUDIO") {
      console.log(`      Channels      : ${track.audioChannels}`);
      console.log(`      Sample Rate   : ${track.sampleRate} Hz`);
    }
    console.log(`      Duration      : ${track.durationSeconds.toFixed(2)}s`);
  }

  // Also test probing from a file
  const tmpFile = path.join(import.meta.dir, "temp_init.mp4");
  fs.writeFileSync(tmpFile, initBuf);
  try {
    console.log(`\n[2] Probing from local file path with Mp4DecryptSession.probe("${tmpFile}")...`);
    const fileInfo = Mp4DecryptSession.probe(tmpFile);
    console.log(`  Verified file probe: ${fileInfo.trackCount} track(s), encrypted: ${fileInfo.tracks[0]?.isEncrypted}`);
  } finally {
    if (fs.existsSync(tmpFile)) fs.unlinkSync(tmpFile);
  }

  console.log("\nMedia probing demo completed successfully!");
}

main().catch(console.error);
